manifests/social/slack.pp

```diff
@@ -0,0 +1,42 @@
+# slack.pp
+# Install Slack for OS X, Ubuntu, or Windows
+# https://www.slack.com, https://packagecloud.io/slacktechnologies
+#
+
+class software::social::slack (
+  $ensure = $software::params::software_ensure,
+) inherits software::params {
+
+  validate_string($ensure)
+
+  case $::operatingsystem {
+    'Ubuntu': {
+      $apt_source_ensure = $ensure ? {
+        'installed' => present,
+        'latest'    => present,
+        default     => $ensure,
+      }
+
+      include '::apt'
+      apt::source { 'slack':
+        ensure   => $apt_source_ensure,
+        comment  => 'Slack APT sources for any Debian-based distribution, including Ubuntu',
+        location => 'https://packagecloud.io/slacktechnologies/slack/debian',
+        release  => 'jessie',
+        repos    => 'main',
+        key      => {
+          'id'     => 'C6ABDCF64DB9A0B2',
+          'source' => 'https://packagecloud.io/slacktechnologies/slack/gpgkey',
+        },
+      }
```

Comment:
I think you missed this require. Its pretty important.
https://github.com/comsolit/puppet-software/blob/d48a2ad251652668d9afc2bda1ccdd21045f44cf/manifests/browsers/chrome.pp#L46

It ensures that the source is added before the apt update is run and also does all that before the package is installed. Notice the arrows on both sides...

Reply:
No, I intentionally left it out, because an `apt-get update` happens automatically.

See the [documentation on apt::source](https://forge.puppet.com/puppetlabs/apt#defined-type-aptsource) on Puppet Forge. It says:
> `notify_update:` _Optional._ Specifies whether to trigger an `apt-get update` run. Valid options: 'true' and 'false'. Default: 'true'.

Reply:
... and for the arrows: I can add an arrow, but I think I tested it, and it works fine just as it is.

Reply:
Let me do some double checking. It used to be necessary with past versions of the apt module, b/c I was bitten by it in the past. They may have fixed it in newer versions. In that case we need to pin our dependency to the newer version or add it to support older apt module versions.

Reply:
I'd feel safer if there was an arrow for the package. I'm pretty sure there is nothing in the apt module to ensure all packages come after the apt source or update. Puppet can run things in different order on different setups, if you don't specify, so a successful test does not guarantee it will work everywhere.

Reply:
Our minimum version of the apt module is 2.1.1 https://github.com/edestecd/puppet-software/blob/master/metadata.json#L14.

That version does not have the notify_update feature. So if we are to stay with that then we need to do it ourselves. Otherwise you need to bump the requirement to the minimum version of the apt module that supports that feature.

Reply:
Looks like the `notify_update` feature was introduced in 2.3.0

2.1.1: https://github.com/puppetlabs/puppetlabs-apt/blob/2.1.1/manifests/source.pp
2.3.0: https://github.com/puppetlabs/puppetlabs-apt/blob/2.3.0/manifests/source.pp

Reply:
Okay, so I add the `-> Class['apt::update'] ->` just now.

Are we then ready to merge?

Reply:
hmm I don't see the change yet. Did you push?

Reply:
Bumping the dependency version also works. Your choice.

Reply:
Sorry, got disturbed. :grimacing: